src/components/ProductsList/ProductsList.tsx

```diff
@@ -1,28 +1,115 @@
-        </ul>
-    </>
-);
+const ProductsList: FC<ProductsListProps> = ({ products, onAddProductToCart }) => {
+    const [searchQuery, setSearchQuery] = useState('');
+    const [selectedFiltersByCategory, setSelectedFilterByCategory] = useState<ProductCategoryName[]>([]);
+    const [selectedSortOption, setSelectedSortOption] = useState<SortOption>(SortOption.PRICE_HIGH_LOW);
+    const [currentPage, setCurrentPage] = useState(1);
+
+    const searchInputReference = useRef<HTMLInputElement>(null);
+
+    const PRODUCTS_PER_PAGE = 8;
+
+    const onSearchButtonClick = () => {
+        if (searchInputReference.current) {
+            setSearchQuery(searchInputReference.current.value);
+            setCurrentPage(1);
+        }
+    };
+
+    const onFilterByCategoryClick = (filterOption: ProductCategoryName) => {
+        const categoryIndex = selectedFiltersByCategory.indexOf(filterOption);
+        if (categoryIndex === -1) {
+            setSelectedFilterByCategory([...selectedFiltersByCategory, filterOption]);
+        } else {
+            const updatedFilters = [...selectedFiltersByCategory];
+            updatedFilters.splice(categoryIndex, 1);
+            setSelectedFilterByCategory(updatedFilters);
+        }
+        setCurrentPage(1);
+    };
+
+    const onSortOptionChange: SortOptionChangeHandler = (sortOption) => {
+        setSelectedSortOption(sortOption);
+        setCurrentPage(1);
+    };
+
+    const onPageChange = (page: number) => {
+        setCurrentPage(page);
+    };
+
+    const filteredAndSortedProducts = useMemo(() => {
+        let filteredProducts = products;
+
+        if (searchQuery.trim() !== '') {
+            filteredProducts = filteredProducts.filter((product) => product.title.toLowerCase().includes(searchQuery.toLowerCase()));
+        }
+
+        if (selectedFiltersByCategory.length > 0) {
+            filteredProducts = filteredProducts.filter((product) => selectedFiltersByCategory.includes(product.category.name));
+        }
+
+        const sortFunction = sortFunctions[selectedSortOption];
+        return [...filteredProducts].sort(sortFunction);
+    }, [products, searchQuery, selectedSortOption, selectedFiltersByCategory]);
+
+    const displayedProducts = useMemo(() => {
+        const startIndex = (currentPage - 1) * PRODUCTS_PER_PAGE;
+        return filteredAndSortedProducts.slice(startIndex, startIndex + PRODUCTS_PER_PAGE);
+    }, [filteredAndSortedProducts, currentPage]);
+
+    return (
+        <>
+            <ControlPanel
+                selectedSortOption={selectedSortOption}
+                onSortOptionChange={onSortOptionChange}
+                products={products}
+                selectedFiltersByCategory={selectedFiltersByCategory}
+                onFilterByCategoryClick={onFilterByCategoryClick}
+                onSearchBtnClick={onSearchButtonClick}
+                searchInputRef={searchInputReference}
+            />
+
+            <ul className={styles.cards}>
```

Comment:
Would be good to add some text here.
And remove pagination
![image](https://github.com/snischuk/fe-react-2024/assets/98031768/b99eb2ec-6593-4e3f-9c22-7eec705676dd)

Reply:
You can fix it in the next pr